spec/lib/msf/core/auxiliary/rocketmq_spec.rb

```diff
@@ -0,0 +1,67 @@
+require 'spec_helper'
+
+RSpec.describe Msf::Auxiliary::Rocketmq do
+  subject do
+    mod = Msf::Auxiliary.new
+    mod.extend(Msf::Exploit::Remote::Tcp)
+    mod.extend(Msf::Auxiliary::Rocketmq)
+    mod
+  end
+
+  describe 'get_rocketmq_version' do
+    context 'correctly looks up id 401 as V4.9.4' do
+      it 'returns that version' do
+        expect(subject.get_rocketmq_version(401)).to eql('V4.9.4')
+      end
+    end
+
+    context 'correctly looks up id 99999 as UNKNOWN.VERSION.ID.99999' do
+      it 'returns that version' do
+        expect(subject.get_rocketmq_version(99999)).to eql('UNKNOWN.VERSION.ID.99999')
+      end
+    end
+  end
+
+  let(:expected_name_server_request) do
+    "\x00\x00\x00\xC7\x00\x00\x00\xC3{\"code\":105,\"extFields\":{\"Signature\":\"/u5P/wZUbhjanu4LM/UzEdo2u2I=\",\"topic\":\"TBW102\",\"AccessKey\":\"rocketmq2\"},\"flag\":0,\"language\":\"JAVA\",\"opaque\":1,\"serializeTypeCurrentRPC\":\"JSON\",\"version\":401}".b
+  end
+
+  let(:expected_name_server_response) do
+    "\x00\x00\x01a\x00\x00\x00_{\"code\":0,\"flag\":1,\"language\":\"JAVA\",\"opaque\":1,\"serializeTypeCurrentRPC\":\"JSON\",\"version\":403}{\"brokerDatas\":[{\"brokerAddrs\":{\"0\":\"172.16.199.135:10911\"},\"brokerName\":\"DESKTOP-8ATHH6O\",\"cluster\":\"DefaultCluster\"}],\"filterServerTable\":{},\"queueDatas\":[{\"brokerName\":\"DESKTOP-8ATHH6O\",\"perm\":7,\"readQueueNums\":8,\"topicSysFlag\":0,\"writeQueueNums\":8}]}".b
+  end
+
+  let(:expected_parsed_data_response) do
+    {"brokerDatas" => [{"brokerAddrs"=>{"0"=>"172.16.199.135:10911"}, "brokerName"=>"DESKTOP-8ATHH6O", "cluster"=>"DefaultCluster"}],
+    "version" => "V4.9.5"}
+  end
+
+  let(:mock_sock) { double :'Rex::Socket::Tcp', send: expected_name_server_request, recv: expected_name_server_response, close: nil, shutdown: nil }
```

Comment:
The `send` result here is wrong:

```
>> sock.send('hello', 0)
=> 5
```

Either this should be updated to `nil` since the result doesn't matter, or the length of the bytes written

```suggestion
  let(:mock_sock) { double :'Rex::Socket::Tcp', send: nil, recv: expected_name_server_response, close: nil, shutdown: nil }
```

or

```suggestion
  let(:mock_sock) { double :'Rex::Socket::Tcp', send: expected_name_server_request.length, recv: expected_name_server_response, close: nil, shutdown: nil }
```

You can also have a smarter mock that returns the byte value programatically:

```suggestion
  let(:mock_sock) do
    mock = double :'Rex::Socket::Tcp',
           recv: expected_name_server_response,
           close: nil,
           shutdown: nil
    allow(mock).to receive(:send) { |req, _timeout| req.length }
    mock
  end
```

But I would choose one of the first two options, most likely just `nil`

Reply:
Thanks for the detailed explantation here Alan, I appreciate it!